state/store.go

```diff
@@ -298,94 +325,97 @@ func (store dbStore) pruneConsensusParams(retainHeight int64) error {
 	// is dependent on.
 	if paramsInfo.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
+		// sanity check that the consensus params at the last height it was changed is there
 		lastRecordedConsensusParams, err := store.loadConsensusParamsInfo(paramsInfo.LastHeightChanged)
 		if err != nil || lastRecordedConsensusParams.ConsensusParams.Equal(&tmproto.ConsensusParams{}) {
 			return fmt.Errorf(
-				"couldn't find consensus params at height %d as last changed from height %d: %w",
+				"couldn't find consensus params at height %d (height %d was originally requested): %w",
 				paramsInfo.LastHeightChanged,
 				retainHeight,
 				err,
 			)
 		}
+
+		// prune the params above the height with which it last changed and below the retain height.
+		err = store.pruneRange(
+			consensusParamsKey(paramsInfo.LastHeightChanged+1),
+			consensusParamsKey(retainHeight),
+		)
+		if err != nil {
+			return err
+		}
 	}
 
-	// batch delete all the consensus params up to the retain height
-	return store.batchDelete(
+	// prune all the consensus params up to either the last height the params changed or if the params
+	// last changed at the retain height, then up to the retain height.
+	return store.pruneRange(
 		consensusParamsKey(1),
-		consensusParamsKey(retainHeight),
 		consensusParamsKey(paramsInfo.LastHeightChanged),
 	)
 }
 
 // pruneABCIResponses calls a reverse iterator from base height to retain height batch deleting
 // all abci responses in between
 func (store dbStore) pruneABCIResponses(height int64) error {
-	return store.batchDelete(abciResponsesKey(1), abciResponsesKey(height), nil)
+	return store.pruneRange(abciResponsesKey(1), abciResponsesKey(height))
 }
 
-// batchDelete is a generic function for deleting a range of keys in reverse order. It will
-// skip keys that have been
-func (store dbStore) batchDelete(start []byte, end []byte, exception []byte) error {
-	iter, err := store.db.ReverseIterator(start, end)
-	if err != nil {
-		return fmt.Errorf("iterator error: %w", err)
-	}
-	defer iter.Close()
-
+// pruneRange is a generic function for deleting a range of keys in reverse order.
+func (store dbStore) pruneRange(start []byte, end []byte) error {
+	var err error
 	batch := store.db.NewBatch()
 	defer batch.Close()
 
-	pruned := 0
-	for iter.Valid() {
-		key := iter.Key()
-		if bytes.Equal(key, exception) {
-			iter.Next()
-			continue
+	// we keep filling up batches of at most 1000 keys, perform a deletion and continue until
+	// we have gone through all of keys in the range. This avoids doing any writes whilst
+	// iterating.
+	for {
+		end, err = store.reverseBatchDelete(batch, start, end)
+		if err != nil {
+			return err
 		}
 
-		if err := batch.Delete(key); err != nil {
-			return fmt.Errorf("pruning error at key %X: %w", key, err)
+		// check if we've iterated through all the keys and that this is the final batch
+		if bytes.Equal(start, end) {
```

Comment:
You may want to make this the loop condition, i.e. `for !bytes.Equal(start, end) { ... }`, and move the `WriteSync()` below, as I think it reads a bit better.

Reply:
If I do that, I will end up creating an empty batch at the end of the loop and calling `WriteSync()` on that. Is that okay?

Reply:
That happens with the existing code too, I believe. In any case, the database should be smart enough to optimize an empty batch write as a noop.

Reply:
Well, I suppose you're right that there'll be an additional batch. Could probably solve it by rearranging the logic. Anyway, not a big deal, do whatever you think seems more natural.

Reply:
Okay I have tried for something that I hope is the best of both worlds